src/condition/structured_condition.hpp

```diff
@@ -0,0 +1,260 @@
+
+struct default_argument_retriever {
+    static constexpr bool is_variadic = false;
+    static constexpr bool is_optional = false;
+};
+
+template <typename T> std::optional<T> convert(const ddwaf_object *obj)
+{
+    if constexpr (std::is_same_v<T, decltype(obj)>) {
+        return obj;
+    }
+
+    if constexpr (std::is_same_v<T, std::string_view> || std::is_same_v<T, std::string>) {
+        if (obj->type == DDWAF_OBJ_STRING) {
+            return T{obj->stringValue, static_cast<std::size_t>(obj->nbEntries)};
+        }
+    }
+
+    if constexpr (std::is_same_v<T, uint64_t> || std::is_same_v<T, unsigned>) {
+        if (obj->type == DDWAF_OBJ_UNSIGNED) {
+            return static_cast<T>(obj->uintValue);
+        }
+    }
+
+    if constexpr (std::is_same_v<T, int64_t> || std::is_same_v<T, int>) {
+        if (obj->type == DDWAF_OBJ_SIGNED) {
+            return static_cast<T>(obj->intValue);
+        }
+    }
+
+    if constexpr (std::is_same_v<T, bool>) {
+        if (obj->type == DDWAF_OBJ_BOOL) {
+            return static_cast<T>(obj->intValue);
+        }
+    }
+
+    return {};
+}
+
+template <typename T> struct argument_retriever : default_argument_retriever {};
+
+template <typename T> struct argument_retriever<unary_argument<T>> : default_argument_retriever {
+    static std::optional<unary_argument<T>> retrieve(const object_store &store,
+        const exclusion::object_set_ref &objects_excluded, const target_definition &target)
+    {
+        auto [object, attr] = store.get_target(target.root);
+        if (object == nullptr || objects_excluded.contains(object)) {
+            return std::nullopt;
+        }
+
+        auto converted = convert<T>(object);
+        if (!converted.has_value()) {
+            return std::nullopt;
+        }
+
+        return unary_argument<T>{target.name, target.key_path,
+            attr == object_store::attribute::ephemeral, std::move(converted.value())};
+    }
+};
+
+template <typename T> struct argument_retriever<optional_argument<T>> : default_argument_retriever {
+    static constexpr bool is_optional = true;
+    static optional_argument<T> retrieve(const object_store &store,
+        const exclusion::object_set_ref &objects_excluded, const target_definition &target)
+    {
+        return argument_retriever<unary_argument<T>>::retrieve(store, objects_excluded, target);
+    }
+};
+
+template <typename T> struct argument_retriever<variadic_argument<T>> : default_argument_retriever {
+    static constexpr bool is_variadic = true;
+    static variadic_argument<T> retrieve(const object_store &store,
+        const exclusion::object_set_ref &objects_excluded,
+        const std::vector<target_definition> &targets)
+    {
+        variadic_argument<T> args;
+        for (const auto &target : targets) {
+            auto arg =
+                argument_retriever<unary_argument<T>>::retrieve(store, objects_excluded, target);
+            if (!arg.has_value()) {
+                continue;
```

Comment:
generally you would want to ensure that there is at least one argument for variadic parameters, no? I have trouble thinking of a condition where that wouldn't be the case, and there it would be better to have the concept of an optional variadic parameter

Reply:
That is checked on line 216, after retrieval.